apps/passport-client/src/service-worker.ts

```diff
@@ -1,85 +1,307 @@
+  // Clean up any stray service workers from prior development.
+  if (!SERVICE_WORKER_ENABLED) {
+    swLog.I(`self-unregistering`);
+    event.waitUntil(
+      (async () => {
+        await swSelf.registration.unregister();
+        for (const client of await swSelf.clients.matchAll()) {
+          (client as WindowClient).navigate(client.url);
+        }
+      })()
+    );
+    return;
+  }
+
+  swLog.V(`activating ${process.env.SW_ID}`);
+
+  event.waitUntil(
+    (async () => {
+      // If supported, NavigationPreload allows fetches to start in
+      // parallel to the running of our fetch event handler.
+      if (swSelf.registration.navigationPreload) {
+        swLog.V(`navigation preload enabled`);
+        await swSelf.registration.navigationPreload.enable();
+      }
+
+      // If cache name has changed, delete any old caches which may exist.
+      await deleteOldCaches();
+
+      // By default, client windows won't be associated with a ServiceWorker
+      // which wasn't already active when they loaded.  This call bypasses that
+      // letting us take over any existing windows.
+      await swSelf.clients.claim();
+
+      swLog.I(`activated ${process.env.SW_ID}`);
+    })()
+  );
+});
+
+/**
+ * Fetch is where the service worker gets to decide what happens to network
+ * requests.  It handles all fetches from controlled pages, even if they're
+ * not within the service worker's origin.
+ */
+self.addEventListener("fetch", (event: FetchEvent) => {
+  if (!shouldHandleFetch(event.request)) {
+    swLog.V(`ignoring fetch ${event.request?.url}`);
+    return;
+  }
+
+  swLog.V(`fetching ${event.request?.url}`);
+
+  //TODO(artwyman): Investigate what respondWidth(undefined) does.
+
+  event.respondWith(
+    (async () => {
+      // Check stable cache first.  If we get a hit, we never fetch.
+      const stableResp = await fetchFromCache(STABLE_CACHE_NAME, event.request);
+      if (stableResp) {
+        swLog.V(`cache hit fetching ${event.request?.url}`);
+        return stableResp;
+      }
+
+      // Next go to the network, and update ephemeral cache with the results.
+      // We only try to look up a result in ephemeral cache when network fails.
+      try {
+        // TODO(artwyman): Add a timeout so we can fall back to cache when slow.
+        // TODO(artwyman): check navigator.onLine to fall back faster.  Docs
+        // suggest it's reliable when false.
+        let resp: Response | undefined = await event.preloadResponse;
+        if (resp) {
+          swLog.V(`preload response fetching ${event.request?.url}`);
+        } else {
+          resp = await fetch(event.request);
+          swLog.V(`network response fetching ${event.request?.url}`);
+        }
+
+        // Response can be read only once, so needs to be cloned to be added
+        // to cache, which happens asynchronously after we return a response.
+        event.waitUntil(
+          updateCache(EPHEMERAL_CACHE_NAME, event.request, resp.clone())
```

Comment:
`updateCache` should probably be named `maybeUpdateCache` - i was a bit scared reading this code. 

another thing that I think would be worth explicitly noting (and this is my understanding that may be incorrect) is that this ephemeral caching is updated only for requests to the CLIENT (not the zupass server), meaning this is a sort of catch-all cache for static resources that we did not have the foresight to add to the permanent cache.